cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,227 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package upgrader
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+	daisyCompute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	"google.golang.org/api/option"
+)
+
+// Parameter key shared with external packages
+const (
+	ClientIDFlagKey = "client_id"
+	DefaultTimeout  = "90m"
+)
+
+const (
+	logPrefix = "[windows-upgrade]"
+
+	metadataKeyWindowsStartupScriptURL       = "windows-startup-script-url"
+	metadataKeyWindowsStartupScriptURLBackup = "windows-startup-script-url-backup"
+
+	versionWindows2008r2 = "windows-2008r2"
+	versionWindows2012r2 = "windows-2012r2"
+)
+
+var (
+	supportedSourceOSVersions = map[string]string{versionWindows2008r2: versionWindows2012r2}
+	supportedTargetOSVersions = reverseMap(supportedSourceOSVersions)
+
+	upgradeScriptName = map[string]string{versionWindows2008r2: "upgrade_script_2008r2_to_2012r2.ps1"}
+
+	expectedCurrentLicense = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2008-r2-dc"}
+	licenseToAdd           = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2012-r2-dc-in-place-upgrade"}
+)
+
+type derivedVars struct {
+	project string
+	zone    string
+
+	osDiskURI        string
+	osDiskType       string
+	osDiskDeviceName string
+	osDiskAutoDelete bool
+
+	instanceName           string
+	machineImageBackupName string
+	osDiskSnapshotName     string
+	newOSDiskName          string
+	installMediaDiskName   string
+
+	windowsStartupScriptURLBackup       *string
+	windowsStartupScriptURLBackupExists bool
+}
+
+// Upgrader implements upgrading logic.
+type Upgrader struct {
```

Comment:
Why is this exported? (naivly I'd expect to see the interface exported, but not the implementation. That said, I don't think you need either of the symbols outside your upgrader package.)

Reply:
It contains input params and should be referenced by main func. Let me extract those params to a separate UpgraderInputParams, and avoid exporting the others.
The reason to do this is make upgrader as a lib, and separate main func from the core.